### PR TITLE
Refactor MessageSet and Message into LegacyRecordBatch 

### PR DESCRIPTION
**Pull request by tvoinarovskyi**
To later support v2 message format

### CODE REVIEW

**Comment by dpkp** [comment]
Looks like there's an error when using pypy:
```
Traceback (most recent call last):
  File "/home/travis/build/dpkp/kafka-python/kafka/producer/sender.py", line 60, in run
    self.run_once()
  File "/home/travis/build/dpkp/kafka-python/kafka/producer/sender.py", line 116, in run_once
    self._metadata, ready_nodes, self.config['max_request_size'])
  File "/home/travis/build/dpkp/kafka-python/kafka/producer/record_accumulator.py", line 502, in drain
    batch.records.close()
  File "/home/travis/build/dpkp/kafka-python/kafka/record/memory_records.py", line 151, in close
    self._buffer = bytes(self._builder.build())
  File "/home/travis/build/dpkp/kafka-python/kafka/record/legacy_records.py", line 440, in build
    self._maybe_compress()
  File "/home/travis/build/dpkp/kafka-python/kafka/record/legacy_records.py", line 418, in _maybe_compress
    compressed = snappy_encode(self._buffer)
  File "/home/travis/build/dpkp/kafka-python/kafka/codec.py", line 136, in snappy_encode
    block = snappy.compress(chunk)
  File "/home/travis/build/dpkp/kafka-python/.tox/pypy/site-packages/snappy.py", line 84, in compress
    return _compress(data)
  File "/home/travis/build/dpkp/kafka-python/.tox/pypy/site-packages/snappy_cffi.py", line 193, in compress
    _input_data, _input_size = prepare(data)
  File "/home/travis/build/dpkp/kafka-python/.tox/pypy/site-packages/snappy_cffi.py", line 183, in prepare
    _out_data = ffi.new('char[]', data)
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/cffi/api.py", line 250, in new
    return self._backend.newp(cdecl, init)
TypeError: 'bytearray' object cannot be interpreted as an index
```

**Comment by tvoinarovskyi** [comment]
The performance on cPython 2.7.10 is better after this PR. On my MacBook (2.7 GHz Intel Core i7) on OSX it yields:

Producer performance (before refactor, `master`):
```
8988.24744641 records/sec (1205293.60883 B/sec), 2.20609148307 latency, 134.0 record size, 5555.60516605 batch size, 41.4298892989 records/req
8979.79654646 records/sec (1204184.38615 B/sec), 2.16188681334 latency, 134.0 record size, 5404.19398294 batch size, 40.2999550965 records/req
8862.39114729 records/sec (1188477.99544 B/sec), 2.17066042747 latency, 134.0 record size, 5181.86904005 batch size, 38.6408137317 records/req
```
Producer performance (after refactor, `legacy_records_refactor` branch):
```
11156.8740853 records/sec (1495023.20879 B/sec), 1.95941633879 latency, 134.0 record size, 5975.13554578 batch size, 44.5905637745 records/req
11154.633705 records/sec (1494721.64645 B/sec), 2.01865647889 latency, 134.0 record size, 6027.51651894 batch size, 44.9516324063 records/req
11110.6124732 records/sec (1488822.26352 B/sec), 1.98895959568 latency, 134.0 record size, 6014.45409253 batch size, 44.8839857651 records/req
```

Consumer performance (before refactor, `master`):
```
25696.5338939 records/sec (3134973.52202 B/sec), 191.90095365 latency, 3.28401605057 fetch/s, 954650.0 fetch size, 1616576.0 max record lag, 7825.0 records/req
25670.0861078 records/sec (3131746.94506 B/sec), 189.899698381 latency, 3.28070306302 fetch/s, 954650.0 fetch size, 1483551.0 max record lag, 7825.0 records/req
25129.3707579 records/sec (3065782.302 B/sec), 189.820181939 latency, 3.21152937427 fetch/s, 954650.0 fetch size, 1358351.0 max record lag, 7825.0 records/req
```
Consumer performance (after refactor, `legacy_records_refactor` branch):
```
32384.5603453 records/sec (4339529.05435 B/sec), 11.7965965736 latency, 4.13878035169 fetch/s, 1048550.0 fetch size, 663526.0 max record lag, 7825.0 records/req
32560.4913518 records/sec (4363110.27358 B/sec), 11.7553327142 latency, 4.16130993247 fetch/s, 1048550.0 fetch size, 499201.0 max record lag, 7825.0 records/req
33022.7197667 records/sec (4425042.82047 B/sec), 11.722275189 latency, 4.22039586671 fetch/s, 1048550.0 fetch size, 334876.0 max record lag, 7825.0 records/req
```

**Comment by tvoinarovskyi** [comment]
So on cPython around 24% on Produce and 28% on Consume (I hope I'm correct =))

**Comment by tvoinarovskyi** [comment]
@dpkp Lol, compared my results to those in https://github.com/dpkp/kafka-python/pull/1185#issuecomment-334266304. Seems like we have similar PC's )

**Comment by tvoinarovskyi** [comment]
@dpkp Rebased on top of deferred parsing, quite nice. One last look and I want to merge it.

File: kafka/record/abc.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def append(self, offset, timestamp, key, value):
+        """ Writes record to internal buffer.
+
+        Arguments:
+            offset (int): Relative offset of record, starting from 0
+            timestamp (int or None): Timestamp in milliseconds since beginning
+                of the epoch (midnight Jan 1, 1970 (UTC)). If omited, will be
```
Comment (jeffwidman) [inline kafka/record/abc.py]: omited --> omitted

File: benchmarks/record_batch_compose.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+from __future__ import print_function
+import perf
+from kafka.record.memory_records import MemoryRecordsBuilder
+import itertools
+import random
+import hashlib
+import os
```
Comment (jeffwidman) [inline benchmarks/record_batch_compose.py]: nit: imports aren't pep8 ordered/spaced

File: benchmarks/record_batch_compose.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+DEFAULT_BATCH_SIZE = 1600 * 1024
+KEY_SIZE = 6
+VALUE_SIZE = 60
+TIMESTAMP_RANGE = [1505824130000, 1505824140000]
+
+# With values above v1 record is 100 bytes, so 10_000 bytes for 100 messages
```
Comment (jeffwidman) [inline benchmarks/record_batch_compose.py]: `10,000 bytes`?

File: benchmarks/record_batch_read.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+from __future__ import print_function
+import perf
+from kafka.record.memory_records import MemoryRecords, MemoryRecordsBuilder
+import itertools
+import random
+import hashlib
+import os
```
Comment (jeffwidman) [inline benchmarks/record_batch_read.py]: imports non-pep8

**Comment by jeffwidman** [comment]
Also, since this isn't merged yet, you might want to fix typo in commit name: `Fix snappy comprassion on PyPy`... comprassion --> compression

File: kafka/producer/record_accumulator.py [hunk 5]
```diff
@@ -174,28 +176,41 @@ class RecordAccumulator(object):
...
     def __init__(self, **configs):
         self.config = copy.copy(self.DEFAULT_CONFIG)
         for key in self.config:
             if key in configs:
                 self.config[key] = configs.pop(key)
 
+        # Convert compression type to INT presentation. Mostly for unit tests,
+        # as Producer should pass already converted values.
+        ct = self.config["compression_type"]
+        self.config["compression_type"] = self._COMPRESSORS.get(ct, ct)
```
Comment (dpkp) [inline kafka/producer/record_accumulator.py]: It's a little confusing to reuse the same configuration name here. It means we have to remember that when talking to users they need this to be a string, but when talking to developers / writing tests / etc, they may need this to be an int. Can we use an internal config like `_compression_type_int` or `compression_type_int` ?

File: kafka/protocol/message.py [hunk 6]
```diff
@@ -154,20 +154,24 @@ class MessageSet(AbstractType):
...
 
     @classmethod
-    def encode(cls, items):
+    def encode(cls, items, prepend_size=True):
         # RecordAccumulator encodes messagesets internally
         if isinstance(items, (io.BytesIO, KafkaBytes)):
             size = Int32.decode(items)
-            # rewind and return all the bytes
-            items.seek(items.tell() - 4)
+            if prepend_size:
```
Comment (dpkp) [inline kafka/protocol/message.py]: I think you also need to alter the number of bytes read depending on whether prepend_size is True / False. Below it is reading size + 4, but that is only correct if we are prepending size. If we are not, I think we only want to read size bytes?

File: kafka/record/legacy_records.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+        self._offset = offset
+        self._timestamp = timestamp
+        self._timestamp_type = timestamp_type
+        self._key = key
+        self._value = value
+        self._crc = crc
+
+    @property
+    def offset(self):
+        return self._offset
```
Comment (dpkp) [inline kafka/record/legacy_records.py]: I'm not familiar with this pattern of using @property to get private attributes. Is that primarily to "enforce" read-only behavior ?

**Comment by dpkp** [comment]
One subtle thing to look out for here is that this alters the output of debug messages from BrokerConnection FetchResponse to include the entire byte buffer for the included messages. In master we do something slightly different. I'm not sure which is better / worse, but we should think about whether we need to adjust this behavior at all.

File: kafka/record/legacy_records.py (hunk 7 above)
Comment (tvoinarovskyi) [reply kafka/record/legacy_records.py]: Yes, it's to enforce read only behaviour

File: kafka/protocol/message.py (hunk 6 above)
Comment (tvoinarovskyi) [reply kafka/protocol/message.py]: Yup, seems like I missed it...

File: benchmarks/record_batch_compose.py (hunk 3 above)
Comment (tvoinarovskyi) [reply benchmarks/record_batch_compose.py]: Yes, it's a habit from developing under 3.6. It allows to write`x = 10_000_000` and will just ignore underscores.

**Comment by tvoinarovskyi** [comment]
@dpkp It's the debug log, but I don't think it needs to show the whole body. How do you look on altering the `Bytes.repr` to limit the amount of data it prints to say 100 bytes?

**Comment by dpkp** [comment]
Showing only first 100 bytes for Bytes object sounds good to me.

**Comment by tvoinarovskyi** [comment]
I'm merging, should have fixed all remaining issues. If I missed something will address it in a separate one.

**Pull request merged**